domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt

```diff
@@ -143,15 +203,247 @@ class TopicListController @Inject constructor(
+              lastCompletedChapterProgress.explorationId != story.chapterList.last().explorationId
+            ) {
+              val lastChapterSummary: ChapterSummary? = story.chapterList.find { chapterSummary ->
+                lastCompletedChapterProgress.explorationId == chapterSummary.explorationId
+              }
+              val nextChapterIndex = story.chapterList.indexOf(lastChapterSummary) + 1
+              val nextChapterSummary: ChapterSummary? = story.chapterList[nextChapterIndex]
+              if (nextChapterSummary != null) {
+                val numberOfDaysPassed =
+                  (Date().time - lastCompletedChapterProgress.lastPlayedTimestamp) / ONE_DAY_IN_MS
+                val promotedStory = createPromotedStory(
+                  storyId,
+                  topic,
+                  completedChapterProgressList.size,
+                  story.chapterCount,
+                  nextChapterSummary.name,
+                  nextChapterSummary.explorationId
+                )
+                if (numberOfDaysPassed < ONE_WEEK_IN_DAYS) {
+                  recommendedStoryBuilder.addRecentlyPlayedStory(promotedStory)
+                } else {
+                  recommendedStoryBuilder.addOlderPlayedStory(promotedStory)
+                }
+              }
+            }
+          }
+          recommendedActivityListBuilder.setRecommendedStoryList(recommendedStoryBuilder)
+        }
+        if (recommendedStoryBuilder.recentlyPlayedStoryCount == 0 && recommendedStoryBuilder.olderPlayedStoryCount > 0
+          && recommendedStoryBuilder.suggestStoryCount == 0
+        ) {
+          recommendedActivityListBuilder.setRecommendedStoryList(recommendedStoryBuilder)
+        }
+        if (recommendedStoryBuilder.recentlyPlayedStoryCount == 0 && recommendedStoryBuilder.olderPlayedStoryCount == 0) {
+          recommendedStoryBuilder.addAllSuggestStory(
+            createRecommendedStoryList(
+              topicProgressList
+            )
+          )
+          recommendedActivityListBuilder.setRecommendedStoryList(recommendedStoryBuilder)
+
+          if (recommendedStoryBuilder.suggestStoryCount == 0) {
+            recommendedActivityListBuilder.setComingSoonTopicList(createComingSoonTopicList())
+            Log.d("topic =", " coming" + " " + "soon")
+          }
+        }
+      }
+    }
+    return recommendedActivityListBuilder.build()
+  }
+
+  private fun createRecommendedStoryList(
+    topicProgressList: List<TopicProgress>
+  ): List<PromotedStory> {
+    val recommendedStories = ArrayList<PromotedStory>()
     topicProgressList.forEach { topicProgress ->
       val topic = topicController.retrieveTopic(topicProgress.topicId)
       topicProgress.storyProgressMap.values.forEach { storyProgress ->
         val storyId = storyProgress.storyId
-        val story = topicController.retrieveStory(topic.topicId, storyId)
+        var story = topicController.retrieveStory(topic.topicId, storyId)
```

Comment:
Should be val (in general, we shouldn't be using vars unless necessary).

Reply:
Used val instead of var.

Reply:
Done.